src/pkg/golang/compiler.go

```diff
@@ -0,0 +1,250 @@
+	s := strings.Fields(string(v))
+	if len(s) < 1 {
+		return fmt.Errorf(efmt, string(v))
+	}
+
+	compiler := c.Compiler
+	compiler.VersionOutput = string(v)
+	compiler.Identifier = s[0]
+	compiler.Type = CompilerTypeFromString(compiler.Identifier)
+	compiler.IsInit = true
+
+	switch compiler.Type {
+
+	case CompilerGo:
+		if len(s) < 3 {
+			return fmt.Errorf(efmt, string(v))
+		}
+		compiler.Version = s[2]
+		compiler.VersionGo = s[2]
+
+	case CompilerTinygo:
+		// e.g. "tinygo version 0.33.0 darwin/arm64 (using go version go1.22.2 and LLVM version 18.1.2)"
+		if len(s) < 8 {
+			return fmt.Errorf(efmt, string(v))
+		}
+		compiler.Version = s[2]
+		compiler.VersionGo = s[7]
+
+		// Fetch additional go-build-tags from tinygo
+		// package fetch needs correct tags to prune
+		cmd := c.compilerCmd("info", "-json")
+		infov, err := cmd.CombinedOutput()
+		if err != nil {
+			return err
+		}
+		var info map[string]interface{}
+		err = json.Unmarshal(infov, &info)
+		if err != nil {
+			return err
+		}
+
+		// extract unique build tags
+		tags := make(map[string]struct{})
+		for _, tag := range c.BuildTags {
+			tags[tag] = struct{}{}
+		}
+		for _, tag := range info["build_tags"].([]interface{}) {
+			tags[tag.(string)] = struct{}{}
+		}
+		for tag := range tags {
+			c.BuildTags = append(c.BuildTags, tag)
+		}
+
+	case CompilerUnkown:
+		return fmt.Errorf(efmt, string(v))
+	}
+	c.Compiler = compiler
+	return nil
+}
+
+// Version returns the Go version string that runtime.Version would return for
+// the Go compiler in this environ.
+func (c *Environ) Version() (string, error) {
+	if err := c.CompilerInit(); err != nil {
+		return "", err
+	}
+	return c.Compiler.VersionGo, nil
+}
+
+func (c Environ) build(dirPath string, binaryPath string, pattern []string, opts *BuildOpts) error {
+
+	if err := c.CompilerInit(); err != nil {
+		return err
+	}
+
+	args := []string{
+		"-o", binaryPath,
+	}
+
+	if c.GO111MODULE != "off" && len(c.Mod) > 0 {
+		args = append(args, "-mod", string(c.Mod))
+	}
+	if c.InstallSuffix != "" {
+		args = append(args, "-installsuffix", c.Context.InstallSuffix)
+	}
+
+	switch c.Compiler.Type {
+	case CompilerGo:
+
+		// Force rebuilding of packages.
+		args = append(args, "-a")
+
+		if opts == nil || !opts.EnableInlining {
+			// Disable "function inlining" to get a (likely) smaller binary.
+			args = append(args, "-gcflags=all=-l")
+		}
+
+		if opts == nil || !opts.NoStrip {
+			// Strip all symbols, and don't embed a Go build ID to be reproducible.
+			args = append(args, "-ldflags", "-s -w -buildid=")
+		}
+
+		if opts == nil || !opts.NoTrimPath {
+			// Reproducible builds: Trim any GOPATHs out of the executable's
+			// debugging information.
+			//
+			// E.g. Trim /tmp/bb-*/ from /tmp/bb-12345567/src/github.com/...
+			args = append(args, "-trimpath")
+		}
+
+	case CompilerTinygo:
+
+		// TODO: handle force-rebuild of packages (-a to standard go)
+		// TODO: handle EnableInlining
```

Comment:
I assume this is postponed? I am not sure about the options tinygo may offer here either. I will do some research about this, but independent of this PR, it should not be blocked by this. I think we can squeeze out some size reduction here by setting the right [build options](https://github.com/tinygo-org/tinygo/blob/release/compileopts/options.go#L22).

Reply:
I think the issues is more that tinygo already does most of these by default so the existing command line arguments don't map well.